addons/controllers/cpi/cpiconfig_controller.go

```diff
@@ -0,0 +1,154 @@
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterapiutil "sigs.k8s.io/cluster-api/util"
+	clusterapipatchutil "sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	cpiv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/cpi/v1alpha1"
+)
+
+// CPIConfigReconciler reconciles a CPIConfig object
+type CPIConfigReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=cpi.tanzu.vmware.com,resources=cpiconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cpi.tanzu.vmware.com,resources=cpiconfigs/status,verbs=get;update;patch
+
+// Reconcile the CPIConfig CRD
+func (r *CPIConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log = r.Log.WithValues("CPIConfig", req.NamespacedName)
+
+	r.Log.Info("Start reconciliation for CPIConfig")
+
+	// fetch CPIConfig resource
+	cpiConfig := &cpiv1alpha1.CPIConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cpiConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("CPIConfig resource not found")
+			return ctrl.Result{}, nil
+		}
+		r.Log.Error(err, "Unable to fetch CPIConfig resource")
+		return ctrl.Result{}, err
+	}
+
+	cpiConfig = cpiConfig.DeepCopy()
+	cluster, err := r.getOwnerCluster(ctx, cpiConfig)
+	if cluster == nil {
+		return ctrl.Result{}, err // no need to requeue if cluster is not found
+	}
+	if res, err := r.ReconcileCPIConfig(ctx, cpiConfig, cluster); err != nil {
+		r.Log.Error(err, "Failed to reconcile CPIConfig")
+		return res, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// ReconcileCPIConfig reconciles CPIConfig with its owner cluster
+func (r *CPIConfigReconciler) ReconcileCPIConfig(ctx context.Context, cpiConfig *cpiv1alpha1.CPIConfig, cluster *clusterapiv1beta1.Cluster) (_ ctrl.Result, retErr error) {
+	patchHelper, err := clusterapipatchutil.NewHelper(cpiConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// patch CPIConfig before returning the function
+	defer func() {
+		r.Log.Info("Patching CPIConfig")
+		if err := patchHelper.Patch(ctx, cpiConfig); err != nil {
+			r.Log.Error(err, "Error patching CPIConfig")
+			retErr = err
+		}
+		r.Log.Info("Successfully patched CPIConfig")
+	}()
+
+	if !cpiConfig.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil
+	}
+	if err = r.ReconcileCPIConfigNormal(ctx, cpiConfig, cluster); err != nil {
+		r.Log.Error(err, "Error reconciling CPIConfig to create/patch data values secret")
+	} else {
```

Comment:
This else is not needed if you just return in line 89

Reply:
I think it will be needed, since r.Log.Error doesn't return. Am I missing something?

Reply:
I mean if you add a return function here that's a more common pattern

Reply:
will change

Reply:
fixed